### DISCUSSION TITLE
Can't bring local Electron app to login: error 400 on token

### DISCUSSION THREAD

**Discussion by verymilan** [zitadel/zitadel#5777] [Q&A]
Hi there, I am trying to use an already fine working webapp wrapped in an Electron client. For some reason I do not understand, it receives a 400 on the final `POST /oauth/v2/token` call.

Its API-server loggs: `Response: {"error":"invalid_grant","error_description":"redirect_uri does not correspond"}`

Since this is a local app, I have enabled "developer mode" in order to be allowed to handle the `redirect_uri=http://127.0.0.1:45735/auth/openid/[...]` part, but still, no luck.

**Comment by hifabienne** [comment]
Hei @verymilan 
This error message means normally that the configured redirect URL does not match the URL that is sent in the request.
Are you sure that they match? Even in developer mode they do have to match.

**Reply by verymilan** [reply]
Hi, thank you for the reply - i at least think they match: 

<img src="https://user-images.githubusercontent.com/10698140/235418338-402a6a70-5527-428b-b124-5fb0e73f1bae.png" style="max-width:350px;" width=400>

While the request then contains: `redirect_uri=http://127.0.0.1:45735/auth/openid/tchncsauth&response_type[...]  302 71 "http://127.0.0.1:45735/"`

It lets me go through the login routine no problem but back on the application, i get: `POST /oauth/v2/token HTTP/2.0" 400`

**Reply by hifabienne** [reply]
Hm. I think in that case you do not send the same redirect url in the post, because otherwise you would not get trough the login.
Can you have a look at the post you are sending to the token endpoint?

**Reply by verymilan** [reply]
I am not sure if i understand / check correctly: the following headers are present according to the browser console:
```
:authority: todo.tchncs.de
:method: POST
:path: /api/v1/auth/openid/tchncsauth/callback # it also does the callback thing on the website where it works
:scheme: https
accept: application/json, text/plain, */*
accept-encoding: gzip, deflate, br
accept-language: de
content-length: 57
content-type: application/json
origin: http://127.0.0.1:45735
referer: http://127.0.0.1:45735/
sec-ch-ua: "Not:A-Brand";v="99", "Chromium";v="112"
sec-ch-ua-mobile: ?0
sec-ch-ua-platform: "macOS"
sec-fetch-dest: empty
sec-fetch-mode: cors
sec-fetch-site: cross-site
user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) vikunja-desktop/v0.20.5 Chrome/112.0.5615.87 Electron/24.1.2 Safari/537.36
```

The request seems to only carry a code as payload: `{code: "7ZLtyZIGE5t......DxJmkg"}`

**Reply by hifabienne** [reply]
Cab you give us some more infos? What kind of authentication flow do you do? PKCE?
Maybe you can add a screenshot of your console configuration of the application.
And can you also send us a har file? you can send it in the discord chat to me or otherwise to fabienne@zitadel.com

**Reply by verymilan** [reply]
**Configuration**

| key | value |
|--------|--------|
| Application type | Web |
| Response Types | Code |
| Auth method | Basic | 
| Grant types | Authorization Code |
| Refresh token | unset |

**Token settings**

| key | value |
|--------|--------|
| Auth token type | Bearer Token |
| User roles inside ID Token | [x] |
| User info inside ID Token | [x] |
| ClockSkew | unset |

**Additional origins**
because 🤷‍♂️ 

* http://127.0.0.1:45735
* http://localhost:45735
****
Gonna have a read on that har file ⏳

**Reply by hifabienne** [reply]
Hei @verymilan 
I can't really see more with the har file.
What I can tell you is that the redirect url is compared with a string equals, so somehow they do not match.
Maybe you can try different possiblities like with the ip, localhost, with / in the end and without / in the end. I am not sure how it is sent
